src/flownet/network_model/_mitchell_fast.py

```diff
@@ -0,0 +1,179 @@
+        if not np.isclose(hull_factor, 1.0):
+            x_hull, y_hull, z_hull = scale_convex_hull_perforations(
+                perforations, hull_factor
+            )
+            x_min = min(x_hull)
+            x_max = max(x_hull)
+            y_min = min(y_hull)
+            y_max = max(y_hull)
+            z_min = min(z_hull)
+            z_max = max(z_hull)
+        else:
+            x_min = min(x)
+            x_max = max(x)
+            y_min = min(y)
+            y_max = max(y)
+            z_min = min(z)
+            z_max = max(z)
+            x_hull = x
+            y_hull = y
+            z_hull = z
+
+        # Determine the convex hull of the original or linearly scaled perforations
+        if np.all(z == z[0]):
+            # 2D cases
+            perforation_hull = Delaunay(np.column_stack([x_hull, y_hull]))
+        else:
+            # 3D cases
+            perforation_hull = Delaunay(np.column_stack([x_hull, y_hull, z_hull]))
+
+    in_hull = np.asarray([False] * num_candidates)
+    x_candidate = np.zeros(num_candidates)
+    y_candidate = np.zeros(num_candidates)
+    z_candidate = np.zeros(num_candidates)
+
+    # Repeat while not all random points are inside the convex hull
+    while not all(in_hull):
+        # Generate a set of random candidates that will be the new
+        # flow nodes
+        x_candidate_tmp = x_min + np.random.rand(num_candidates) * (x_max - x_min)
+        y_candidate_tmp = y_min + np.random.rand(num_candidates) * (y_max - y_min)
+        z_candidate_tmp = z_min + np.random.rand(num_candidates) * (z_max - z_min)
+
+        # Update the list of flow node candidates. Only the points that previously
+        # were not inside the convex hull are updated.
+        np.putmask(x_candidate, np.invert(in_hull), x_candidate_tmp)
+        np.putmask(y_candidate, np.invert(in_hull), y_candidate_tmp)
+        np.putmask(z_candidate, np.invert(in_hull), z_candidate_tmp)
+
+        candidates = np.vstack([x_candidate, y_candidate, z_candidate]).T
+
+        if concave_hull_bounding_boxes is not None:
+            in_hull = check_in_hull(
+                concave_hull_bounding_boxes, candidates, in_hull_known=in_hull
+            )
+        else:
+            # Test whether all points are inside the convex hull of the perforations
+            if np.all(z == z[0]):
+                in_hull = perforation_hull.find_simplex(candidates[:, (0, 1)]) >= 0
+            else:
+                in_hull = perforation_hull.find_simplex(candidates) >= 0
+
+    # Generate all new flow nodes
+    for i in range(num_points, num_points + num_added_flow_nodes):
+        mid = time.time()
+        if mid - start > 4:
+            start = mid
+            print(
+                f"\rAdding flow nodes:  {int(((i-num_points)/num_added_flow_nodes)*100)}%",
+                end="",
+            )
+
+        delta_x_relative = np.power(
+            (x.repeat(num_candidates) - x_candidate.repeat(x.shape[0]))
```

Comment:
using `repeat `is actually quite neat, I think, but in order to get all combinations correct I think you need to use `tile` also
```suggestion
            (np.tile(x, num_candidates) - x_candidate.repeat(x.shape[0]))
```